### PR TITLE
Refactor the map loader

### PR DESCRIPTION
**Pull request by SpaceManiac**
:cl:
refactor: The map loader has been cleaned up and made more flexible.
/:cl:

Follow-up to #38764. What is hopefully the bulk of the refactor is done, and one tweak to SSmapping has been made to avoid duplicate work, which should cut about 10% off of station and Lavaland load times. Tested station maps, ruins, placing templates, and buildmode reload, and everything seemed to work. **TODO**: test Reebe loading.

* `load_map` is now a global proc
* `/datum/map_reader` is gone
* `/datum/parsed_map/New` parses map files
* `/datum/parsed_map/load` modifies the world accordingly
* The MC critical section for map loading no longer includes parsing

Future features that this work helps enable:
* Validating maps without loading them, a prerequisite for the current loose design for the out-of-tree map system.
* Properly addressing some of the away mission loading issues (#39378) and other atmos issues relating to map templates.
* Allowing cropping by Z as well as X and Y, to improve the behavior of buildmode reload for multi-z maps.
* Tricks such as this fancy preview for the "Map Template - Place" verb (not in this PR because I'm not satisfied with it yet):  \
  \
  advanced-preview-cropped

### CODE REVIEW

**Comment by SpaceManiac** [comment]
Travis failure is spurious

File: code/modules/mapping/reader.dm [hunk 1]
```diff
@@ -10,77 +10,57 @@ GLOBAL_DATUM_INIT(_preloader, /datum/map_preloader, new)
...
-	var/static/regex/dmmRegex = new/regex({""(\[a-zA-Z]+)" = \\(((?:.|\n)*?)\\)\n(?!\t)|\\((\\d+),(\\d+),(\\d+)\\) = \\{"(\[a-zA-Z\n]*)"\\}"}, "g")
-		// /^[\s\n]+"?|"?[\s\n]+$|^"|"$/g
-	var/static/regex/trimQuotesRegex = new/regex({"^\[\\s\n]+"?|"?\[\\s\n]+$|^"|"$"}, "g")
-		// /^[\s\n]+|[\s\n]+$/
-	var/static/regex/trimRegex = new/regex("^\[\\s\n]+|\[\\s\n]+$", "g")
-	#ifdef TESTING
-	var/turfsSkipped
-	#endif
+	var/list/modelCache
+	var/list/bad_paths
```
Comment (AnturK) [inline code/modules/mapping/reader.dm]: You don't output these anywhere ?

Comment (Cyberboss) [reply code/modules/mapping/reader.dm]: Suppose it's fine as a public var, this will come in handy when public maps happen

Comment (SpaceManiac) [reply code/modules/mapping/reader.dm]: `modelCache` is to allow `build_cache()` to be called twice; `bad_paths` is for future use.

**Comment by Nabski89** [comment]
I've had some issues with loading custom templates crashing the server. Is that still a thing here?

**Comment by SpaceManiac** [comment]
I'd need more details.

**Comment by Nabski89** [comment]
image

in this round
https://tgstation13.org/raw-logs/basil/data/logs/2018/08/06/round-92138/

I uploaded a custom map, then when trying to place it it loads the green square then dies.

**Comment by SpaceManiac** [comment]
Do you have the file you were using? Did it get to the "Are you sure" prompt or die before that? It looks like it was highpop (~75 players at roundstart) which could be part of the problem.

**Comment by Nabski89** [comment]
[goal_lowpop - Copy.log](https://github.com/tgstation/tgstation/files/2264460/goal_lowpop.-.Copy.log)
That's really a .dmm that I just renamed as a log. I don't know how much that breaks it.

**Comment by SpaceManiac** [comment]
I successfully reproduced the failure. Being an infinite loop, it seems plausible that highpop conditions could turn the failure to a crash. I'll look into it.

**Comment by SpaceManiac** [comment]
I take that back: I do know why that fixes it.

Your template uses a nonexistent area, `/area/station_goal`. Due to what appears to be a BYOND bug, the `continue` statement hit when a nonexistent path is thrown away skips the condition on the `do`-`while`, causing the loop to never terminate.

**Comment by Nabski89** [comment]
Ohhhhh. That would explain why they USED to work, but then stopped at some point. Thank you.

Oh well, people always enjoy antag tokens.

**Comment by Cyberboss** [comment]
This good @SpaceManiac ?

**Comment by SpaceManiac** [comment]
Should be good to go

**Pull request merged**